### PR TITLE
pmon services and daemons test plan

### PR DESCRIPTION
**Pull request by sujinmkang**
### Description of PR

This is the test plan for verifying all daemons running inside PMON.
Each daemon test will check the DB state to verify its running state on the platform.
Any failure result can provide any regression between images during regression test.

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### CODE REVIEW

**Review by jleveque — CHANGES_REQUESTED** [review]
For the daemons which write info to State DB, we also need to add test cases to ensure that the daemons delete their tables from State DB upon shutdown.

We need to test a few different shutdown methods:
- supervisorctl stop <daemon> (this should send SIGTERM to the process by default)
- Send other signals which the daemon is expected to shut down via (depends on the signal handling of each daemon, but most are programmed to exit on SIGTERM and SIGINT)

In this process, we should confirm that the daemon exits (we can also check for appropriate log messages, and check that all applicable tables are removed from State DB.

We then start the daemon again and ensure that the tables get re-populated.

**Comment by sujinmkang** [comment]
@jleveque It's good point. I will add that. Initially I had stop and start but didn't have the check of DB clean up when a daemon exits.

File: docs/testplan/PMON-Services-Daemons-test-plan.md [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+
+# Test setup
+No setup pre-configuration is required, test will configure and clean-up all the configuration.
+The sanity-check plugin needs to be run before running this test, which will cover the verification of pmon docker running on SONiC system.
+
+## Test configuration information file 
+To check the proper daemon running status along with the db keys and fields associated to the daemon based on the image version, 
+"pmon_daemons_info.json" will provide the exit signals, DB location, tables, keys, and fields information to check based on the daemon version.
+Each daemon needs to support `{daemon} -v` to show the version of daemon.
+If a daemon doesn't support the version command, then the test assums the version as 1.0 (default for all daemons in PMON).
```
Comment (jleveque) [inline docs/testplan/PMON-Services-Daemons-test-plan.md:45]: s/assums/assumes/

File: docs/testplan/PMON-Services-Daemons-test-plan.md [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+
+3. Perform the following steps for the list of the expected daemon
+	a. Verify the daemon is running and check if all expected DB data is populated
+   b. Disable the critical process autorestart if it's enabled using [disable_and_enable_autorestart](https://github.com/Azure/sonic-mgmt/blob/73830827460750d88ba0a5de71bb5c79ad33c46e/tests/process_monitoring/test_critical_process_monitoring.py#L37)
+	c. Stop the daemon using `sudo docker exec pmon supervisorctl stop {daemon}`
+   d. Verify the daemon is not running using `sudo docker exec pmon supervisorctl status {daemon}` and check if all DB data is cleared 
+	e. Restart the daemon using `sudo docker exec pmon supervisorctl start {daemon}`
+   f. Verify the daemon is running using `sudo docker exec pmon supervisorctl status {daemon}` and check if all expected DB data is populated
+   h. Repeat from step c to step f for all exit signals which the daemon can exit with
+   i. Enable the critical process autorestart
```
Comment (jleveque) [inline docs/testplan/PMON-Services-Daemons-test-plan.md:220]: These are common test steps that we want to perform on all State DB-populating daemons (chassisd, pcied, psud, syseepromd, thermalctld, xcvrd). I think you could rework this section so that it describes the common test procedure for every State DB-populating daemon, and in the section below, for those daemons, refer to this procedure and simply specify the data that will be verified for that daemon along with any daemon-specific testing we want to perform outside of this procedure.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: docs/testplan/PMON-Services-Daemons-test-plan.md [hunk 3]
```diff
@@ -136,8 +136,8 @@ If a daemon doesn't support the version command, then the test assums the versio
 
 ## Test case # 1 All expected daemons are running
 ### Test objective
-Verify each daemon status in PMON docker using "check_pmon_daemon_status()" in "tests/common/platform/daemon_utils.py"
-Extend "check_pmon_daemon_status()" if there is any gap to follow the steps below.
+Verify all daemon staus in PMON container using "check_pmon_daemon_status()" in "tests/common/platform/daemon_utils.py"
```
Comment (jleveque) [inline docs/testplan/PMON-Services-Daemons-test-plan.md:139]: s/staus/status

Suggest rewording to "Verify the status of all daemons in PMon container using ..."

File: docs/testplan/PMON-Services-Daemons-test-plan.md [hunk 4]
```diff
@@ -209,25 +209,34 @@ Extend "check_pmon_daemon_status()" if there is any gap to follow the steps belo
...
+4. Verify the daemon is not running using `sudo docker exec pmon supervisorctl status {daemon}` 
+5. Verify all DB data is cleared if any DB data has been populated while daemon was running 
+6. Restart the daemon using `sudo docker exec pmon supervisorctl start {daemon}`
+7. Verify the daemon is running using `sudo docker exec pmon supervisorctl status {daemon}`
+8. Verify all expected DB data is populated same as step 1.
+9. Repeat from step 3 to step 8 with sending the signal to kill the daemon for each exit signal being handled by the daemon
+10. Enable the critical process autorestart
 
-## Test case # 2 Data for chassisd daemon
+## Test case # 3 Data for chassisd daemon
```
Comment (jleveque) [inline docs/testplan/PMON-Services-Daemons-test-plan.md:230]: Rather than calling these "test cases", I think they should simply be called "Daemon-specific database data verification procedures". Then, step 1 above can simply read, `1. Verify the daemon is running and check if all expected DB data is populated (refer to daemon-specific database data verification procedure below)`.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: docs/testplan/PMON-Services-Daemons-test-plan.md (hunk 4 above)
Comment (sujinmkang) [reply docs/testplan/PMON-Services-Daemons-test-plan.md:230]: I removed the branch accidently so please review it from https://github.com/Azure/sonic-mgmt/pull/3213

**Comment by sujinmkang** [comment]
Please review https://github.com/Azure/sonic-mgmt/pull/3213 instead.
